materialize-rockset/rockset.go

```diff
@@ -67,31 +77,98 @@ func getCollection(ctx context.Context, client *client, workspace string, collec
-	ctx, cancel := context.WithTimeout(ctx, time.Minute*3)
-	defer cancel()
-
+// awaitCollectionReady blocks until the given Rockset collection has a READY status AND has completed
+// ingestion of all the files in the cloud storage bucket for the given integration. This is required in order to
+// successfully hand off from a cloud storage materailization while preserving the ingestion order of documents.
+// Basically, we need to wait until all the documents in the cloud storage bucket have been ingested before we can start
+// using the write API, or else an earlier document from cloud storage may overwrite the one we ingest using the write
+// API.
+func awaitCollectionReady(ctx context.Context, client *client, workspace, collectionName, integration string) error {
 	for {
-		select {
-		case <-ctx.Done():
-			return nil, fmt.Errorf("timed out waiting for collection `%s/%s` to be created: %w", workspace, collection.Name, err)
-		default:
-			if col, err := client.GetCollection(ctx, workspace, collection.Name); err != nil {
-				return nil, fmt.Errorf("fetching collection `%s/%s` while waiting for it to be created: %w", workspace, collection.Name, err)
-			} else if col.Status != "READY" {
-				// Not ready yet, so wait a couple of seconds then we'll try again.
-				time.Sleep(time.Second * 2)
-			} else {
-				return col, nil
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		var collection, err = getCollection(ctx, client, workspace, collectionName)
+		if err != nil {
+			return err
+		}
+		if collection == nil {
+			return fmt.Errorf("rockset collection '%s' does not exist or has been deleted", collectionName)
+		}
+
+		var ready = collection.Status == STATUS_READY
+
+		// We'll check to see if a collection has made no progress and emit a specific warning in that case.  If a
+		// Rockset collection isn't using a large enough instance, then it'll silently refuse to actually use the
+		// integration that was configured. This is a really easy mistake to make, and there doesn't appear to be a way
+		// to actually detect the instance size via the API, so we'll try to log a helpful warning if the collection
+		// goes too long without downloading any of the objects from cloud storage.
+		var bulkIngestStalled = false
+		var collectionAge = time.Since(collection.CreatedAt)
+		// If an integration was specified, then we need to _also_ wait for that integration to fully process all of its
+		// files. Technically, this will already be the case if there was enough data in the source bucket to trigger a
+		// "bulk load", because in that case Rockset will not set the status to READY until after all the files have
+		// been uploaded. BUT, if the source bucket didn't contain enough data to trigger the "bulk load", then the
+		// status may be set to READY before all of the source files have been processed. This condition is a guard
+		// against ingesting data out of order in that specific case.
+		if integration != "" {
```

Comment:
If they don't specify an integration, couldn't we return at the top of this function? In that case, there is no bulk ingestion to perform, right?

Reply:
No, because this may still be needed in order to await the `READY` status of a collection that was created without any integration. Previously, this was being done as part of creating the collection during `Apply`, but I ripped it out of there so that this is the only place where we wait for the collection to be ready.